pkg/virt-controller/watch/vmi.go

```diff
@@ -2123,3 +2156,184 @@ func (c *VMIController) getVolumePhaseMessageReason(volume *virtv1.Volume, names
 	}
 	return virtv1.VolumePending, PVCNotReadyReason, "PVC is in phase Lost"
 }
+
+func (c *VMIController) handleInterfaceHotplug(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod) error {
+	log.Log.Object(pod).V(4).Info("network iface hotplug for pod")
+	multusAnnotation, err := services.GenerateMultusCNIAnnotationWithInterfaceNamingScheme(
+		vmi,
+		namescheme.CreateHashedNetworkNamingScheme,
+	)
+	if err != nil {
+		return err
+	}
+
+	currentMultusAnnotation, found := pod.Annotations[networkv1.NetworkAttachmentAnnot]
+	if found {
+		log.Log.Object(pod).V(4).Infof("current multus annotation for pod: %s", currentMultusAnnotation)
+	}
+
+	if multusAnnotation != "" {
+		pod.Annotations[networkv1.NetworkAttachmentAnnot] = multusAnnotation
+		log.Log.Object(pod).V(4).Infof("updated multus annotation for pod: %s", multusAnnotation)
+	} else {
+		pod.Annotations[networkv1.NetworkAttachmentAnnot] = "[]"
+		log.Log.Object(pod).V(4).Info("cleared pod cncf.io/networks annotation")
+	}
+
+	if currentMultusAnnotation != multusAnnotation {
+		pod, err = c.clientset.CoreV1().Pods(pod.GetNamespace()).UpdateStatus(context.Background(), pod, v1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *VMIController) doesPodNeedIfaceHotplug(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod) bool {
+	indexedNetworkStatus := nonDefaultPodMultusNetworksIndexedByIfaceName(pod)
+	for i, iface := range vmispec.FilterMultusNonDefaultNetworks(vmi.Spec.Networks) {
```

Comment:
@maiqueb  I found that when remove interface, pod annotation doesn't reflect the change, and realized that the function `doesPodNeedIfaceHotplug` may not handle it properly.

should it range indexedNetworkStatus to find it is been unplugged? otherwise FilterMultusNonDefaultNetworks may return empty( it then skip unplug annotation update).

for example the following change(to help express what I mean)
```go
 func (c *VMIController) doesPodNeedIfaceHotplug(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod) bool {
        indexedNetworkStatus := nonDefaultPodMultusNetworksIndexedByIfaceName(pod)
-       for i, iface := range vmispec.FilterMultusNonDefaultNetworks(vmi.Spec.Networks) {
-               if _, found := indexedNetworkStatus[namescheme.SecondaryInterfaceName(i+1)]; found {
-                       continue
-               }
-               if _, found := indexedNetworkStatus[namescheme.HashNetworkName(iface.Name)]; !found {
-                       return true
+       networks := vmispec.FilterMultusNonDefaultNetworks(vmi.Spec.Networks)
+
+       deletedFromSpec := true
+       for ifaceName := range indexedNetworkStatus {
+               for _, network := range networks {
+                       if ifaceName == namescheme.HashNetworkName(network.Name) {
+                               deletedFromSpec = false
+                       }
                }
        }
+       if deletedFromSpec {
+               return true
+       }
        return false
 }
```

now it can success trigger pod annotation change( make dynamic-networks-controller to remove interface. ( removed dummy interface though ).

It seems there's no reverse of plugphase1 exist, 
1. I'm not sure the semantic of removeinterface ( keep three interface there is ok? ) 
2. and since cni remove the dummy interface, the host side veth interface will leak too.

Reply:
good to know, thanks.

Reply:
by the way, from the test,  my above function: `handleInterfaceHotplug` change have flaw, that add multiple interface failed.
here's the new version of func: `handleInterfaceHotplug`

```diff
diff --git a/pkg/virt-controller/watch/vmi.go b/pkg/virt-controller/watch/vmi.go
index 6a55430c6..eb2f342a8 100644
--- a/pkg/virt-controller/watch/vmi.go
+++ b/pkg/virt-controller/watch/vmi.go
@@ -2192,7 +2192,9 @@ func (c *VMIController) handleInterfaceHotplug(vmi *virtv1.VirtualMachineInstanc
 
 func (c *VMIController) doesPodNeedIfaceHotplug(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod) bool {
        indexedNetworkStatus := nonDefaultPodMultusNetworksIndexedByIfaceName(pod)
-       for i, iface := range vmispec.FilterMultusNonDefaultNetworks(vmi.Spec.Networks) {
+       networks := vmispec.FilterMultusNonDefaultNetworks(vmi.Spec.Networks)
+
+       for i, iface := range networks {
                if _, found := indexedNetworkStatus[namescheme.SecondaryInterfaceName(i+1)]; found {
                        continue
                }
@@ -2200,6 +2202,18 @@ func (c *VMIController) doesPodNeedIfaceHotplug(vmi *virtv1.VirtualMachineInstan
                        return true
                }
        }
+
+       deletedFromSpec := true
+       for ifaceName := range indexedNetworkStatus {
+               for _, network := range networks {
+                       if ifaceName == namescheme.HashNetworkName(network.Name) {
+                               deletedFromSpec = false
+                       }
+               }
+       }
+       if deletedFromSpec {
+               return true
+       }
        return false
 }
```

Reply:
if you agree, should I create a pull request for these two change on your repo: https://github.com/maiqueb/kubevirt/tree/hotplug-api ?

Reply:
How can you add multiple interfaces in a single request ? This can only be triggered by virtctl (or any other components using the kubevirt service account).

Never the less, I am still re-rewriting this section, accounting for unplug; right now, I think I do not need your code, given the base will change. 

But again, please continue issuing feedback; it will be valuable.

Reply:
> How can you add multiple interfaces in a single request 
I mean execute virtctl addinterface many time,  the pod annotation networks will include multiple attachment interface .

> But again, please continue issuing feedback; it will be valuable.
Got it.

Reply:
> > How can you add multiple interfaces in a single request
> 
> I mean execute virtctl addinterface many time, the pod annotation networks will include multiple attachment interface .

I see. Yeah, this scenario will have to be accounted for. I plan on re-writing these funcs, to provide a better reconciliation loop experience.

> 
> > But again, please continue issuing feedback; it will be valuable.
> 
> Got it.

Thanks again.